### ISSUE TITLE
NDNApp limited to One Thread?

### ISSUE THREAD

**Issue by justincpresley** [named-data/python-ndn#31]
When trying to use an NDNApp access in another thread and prepare data, I receive the following error:

```
 data_packet = self.app.prepare_data(int_name + [Component.from_number(packet_number, Component.TYPE_SEGMENT)], byte_chunk, meta_info=mi)
  File "/home/corbin/.local/lib/python3.8/site-packages/ndn/app.py", line 143, in prepare_data
    signer = self.keychain.get_signer(kwargs)
  File "/home/corbin/.local/lib/python3.8/site-packages/ndn/security/keychain/keychain_sqlite3.py", line 414, in get_signer
    identity = self.default_identity()
  File "/home/corbin/.local/lib/python3.8/site-packages/ndn/security/keychain/keychain_sqlite3.py", line 335, in default_identity
    cursor = self.conn.execute('SELECT id, identity, is_default FROM identities WHERE is_default=1')
sqlite3.ProgrammingError: SQLite objects created in a thread can only be used in that same thread. The object was created in thread id 140249821718336 and this is thread id 140249736873728.
```

Are you unable to use an NDNApp in multiple threads? If not, If I create multiple instances of an 'NDNApp's, will published data from one App satisfy another App's interests?

**Comment by justincpresley** [comment]
I have gotten coroutines to work well with multithreading in the past, but it requires careful coding. It did seem to be the best way to get the best performance too instead of queues. However, I have not heard about the GIL issue though, so that last statement is subject to alter.

I believe the main problem here is the fact that NDNApp is not thread-safe. **Will this be a future improvement?**

So far, I have encountered two applications ([ndn-hydra](https://github.com/UCLA-IRL/ndn-hydra) + [nr-archway](https://github.com/justincpresley/nr-archway)) that could heavily benefit from / require using NDN-related API calls (i.e. python-ndn) in multiple threads (with coroutines). More are bound to be encountered in the future.
I have tried to use queuing for handing off data/commands in one application; however, the performance (considering NDN's current performance) indicts that it was not a viable solution.
